tests/storage/test_room_search.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import synapse.rest.admin
+from synapse.rest.client import login, room
+from synapse.storage.engines import PostgresEngine
+
+from tests.unittest import HomeserverTestCase
+
+
+class NullByteInsertionTest(HomeserverTestCase):
+    servlets = [
+        synapse.rest.admin.register_servlets_for_client_rest_resource,
+        login.register_servlets,
+        room.register_servlets,
+    ]
+
+    def test_null_byte(self):
+        """
+        Postgres/SQLite don't like null bytes going into the search tables. Internally
+        we replace those with a space.
+
+        Ensure this doesn't break anything.
+        """
+
+        # register a user and create a room, create some messages
+        self.register_user("alice", "password")
+        access_token = self.login("alice", "password")
+        room_id = self.helper.create_room_as("alice", True, "1", access_token)
+        body1 = "hi\u0000bob"
+        body2 = "another message"
+        body3 = "hi alice"
+
+        # send messages and ensure they don't cause an internal server
+        # error
+        resp1 = self.helper.send(room_id, body1, "1", access_token)
+        resp2 = self.helper.send(room_id, body2, "2", access_token)
+        resp3 = self.helper.send(room_id, body3, "3", access_token)
+        self.assertTrue("event_id" in resp1)
+        self.assertTrue("event_id" in resp2)
+        self.assertTrue("event_id" in resp3)
+
+        # check that search still works with the message where the null byte was replaced
+        store = self.hs.get_datastore()
+        res1 = self.get_success(
+            store.search_msgs([room_id], "hi bob", ["content.body"])
+        )
+        self.assertEquals(res1.get("count"), 1)
```

Comment:
I think it would be clearer to put the highlights checks here after the request. We won't need to keep as much state in our head as we move down through the test then.

(And we can use a single `response` variable that gets re-used instead of storing them separately).